scripts/detection/ssd/train_ssd.py

```diff
@@ -253,11 +388,25 @@ def train(net, train_data, val_data, eval_metric, ctx, args):
             warnings.simplefilter("always")
             net.initialize()
             async_net.initialize()
+            # needed for net to be first gpu when using AMP
+            net.collect_params().reset_ctx(ctx[0])
```

Comment:
I got higher mAP when I am not using amp and removing this line,
I got lower mAP when I am not using amp and keep this line,
cc @zhreshold @Kh4L

Reply:
How big is the mAP difference?

Reply:
8% mAP in the first epoch with my own dataset. I think it is related to the GPU initialization with the empty data used to obtain the anchors.

I was able to obtain comparable mAP on the first epoch, I had reloaded the entire network after get_dataloader, calling get_model and initialize again.

Reply:
When you leave the line the anchors are first generated by the model on GPU 0 then copied to CPU, but when comment out this line, you are using anchors generated directly the model on the CPU. Both should produce an identical result.

> I was able to obtain comparable mAP on the first epoch, I had reloaded the entire network after get_dataloader, calling get_model and initialize again.

You managed to get a higher mAP with `net.collect_params().reset_ctx(ctx[0])`, but you had to add lines to reload the whole model after `get_dataloader`. Did I understand correctly?

Reply:
Yes,

I managed to get a higher mAP with `net.collect_params().reset_ctx(ctx[0])`, but I had to add lines to reload the whole model after `get_dataloader`.

It is strange, I was expecting to obtain the same results.

Reply:
@PariksheetPinjari909  I think it's weird according to your description. Will you be able to compare the anchors generated w/without removing this line?

Reply:
The anchors are the same, maybe It could be related to the weight initialization.

It looks like the issue occurs when a resnet is used as backbone such as ssd_512_resnet50_v1_custom.

I will try to prepare a code sample.